internal/netTestutils.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os/exec"
+
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+// LocalNetworkServices provides a way of running bare-bones
+// DNS and HTTP servers bound to the ip of a created device,
+// which enables testing VM networking without relying on an
+// external network.
+type LocalNetworkServices interface {
+	// DNSServerIP returns the IP address of the DNS server
+	// (useful for seting resolv.conf)
+	DNSServerIP() string
+
+	// URL returns the full URL to use when querying the provided
+	// subpath. i.e. if subpath is "foo/bar", this will return
+	// something like "http://phony.domain.test/foo/bar"
+	URL(subpath string) string
+
+	// Serve will start the DNS and HTTP servers, blocking until
+	// either of them shuts down.
+	Serve(ctx context.Context) error
+}
+
+// NewLocalNetworkServices returns an implementation of
+// LocalNetworkServices where the DNS server will bind to the
+// ip of a created device and serve a single mapping from
+// "domain" to that ip. The HTTP server will serve a pages at
+// paths defined in the keys of "webpages", with the content set
+// in the values of that map.
+func NewLocalNetworkServices(webpages map[string]string) (LocalNetworkServices, error) {
+	testDevName := "testdev0"
+	ipAddr := "10.0.0.1"
+	ipCidr := fmt.Sprintf("%s/32", ipAddr)
+
+	output, err := exec.Command("ip", "tuntap", "add", testDevName, "mode", "tun").CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to add tun dev, \"ip\" output: %s", string(output))
```

Comment:
nit: you can use \` instead of " for strings where you want to embed more "s.  Also, the `%q` format verb can be useful in quoting/escaping a string:

```
errors.Wrapf(err, `failed to add tun dev, "ip" output: %q`, string(output))
```